torch/nn/_functions/rnn.py

```diff
@@ -8,19 +8,22 @@
 
 
 def RNNReLUCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.relu(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+    x_h = input if w_ih is None else F.linear(input, w_ih, b_ih)
+    hy = F.relu(x_h + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
 def RNNTanhCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.tanh(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+    x_h = input if w_ih is None else F.linear(input, w_ih, b_ih)
+    hy = F.tanh(x_h + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
 def LSTMCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
     hx, cx = hidden
-    gates = F.linear(input, w_ih, b_ih) + F.linear(hx, w_hh, b_hh)
-    ingate, forgetgate, cellgate, outgate = gates.chunk(4, 1)
+    x_h = input.unsqueeze(1).expand(input.size(0), 4, input.size(1)) if w_ih is None else F.linear(input, w_ih, b_ih)
```

Comment:
the modern definition of lstm does indeed use different parameters for each linear projection in each gate. see section 4.6.1 of alex graves' [thesis](https://www.cs.toronto.edu/~graves/preprint.pdf). although an older version did not have connections from the input vector to the gates at all (see figures [here](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.55.5709)), although i don't know of any modern code that uses gers' formulation as is. and fwiw the original formulation didn't have a [forget gate](http://www.bioinf.jku.at/publications/older/2604.pdf)...